WARN: Cannot find the file '...' in module '...' base dir '...', skipping violations.

### Description

If the source code is in subdirectories, the comparison of the paths is case-sensitive. (Also on Windows)

The output of MSBuild contains all paths in lowercase (see buildlog.log in log section). However, the project contains a subdirectory "Sources" in which all source files are located. This means that the files cannot be found.

### Steps to reproduce the problem

```PowerShell
PS > git clone https://github.com/rudolfgrauberger/sonar-cxx-examples.git
PS > cd sonar-cxx-examples/FolderStructureCpp
PS > MSBuild.exe .\FolderStructureCpp.vcxproj /t:rebuild /p:Configuration=Release /fileLogger "/fileLoggerParameters:LogFile=buildlog.log;Verbosity=normal;Encoding=UTF-8"
PS > sonar-scanner
```

### Expected behavior

That the compiler sensor finds the file and analyzes it. Under Windows the upper and lower case "actually" doesn't matter. 

If I have seen this correctly here (#155), then it was already a topic once.  Could it be that it was lost during the last refactoring (#1180)?

### Actual behavior

Because the file is not found, it is not analyzed.

### Known workarounds

Not good, but works...

#### Rename folder to lower case
```PowerShell
PS > mv Sources sources_temp
PS > mv sources_temp sources
```
### LOG file

**stdout**
```
...
INFO: Searching reports by relative path with basedir 'E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp' and search prop 'sonar.cxx.vc.reportPath'
INFO: Parser will parse '1' report file(s)
INFO: Processing report 'E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\buildlog.log'
INFO: Parsing 'Visual C++' initialized with report 'E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\buildlog.log', Charset= 'UTF-8'
INFO: Using pattern : '(.*>)?\x20*(?<file>.*)\((?<line>\d+)\):\x20warning\x20(?<id>C\d+):(?<message>.*)'
WARN: Cannot find the file 'e:\dev\source\repos\sonar-cxx-samples\folderstructurecpp\sources\main.cpp' in module 'FolderStructureCpp' base dir 'E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp', skipping violations.
INFO: CXX-COMPILER-VC processed = 0
...
```

**buildlog.log**
```
Der Buildvorgang wurde am 03.01.2019 01:47:41 gestartet.
Projekt "E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj" auf Knoten "1", rebuild Ziel(e).
_PrepareForClean:
  Die Datei "Release\FolderSt.67F13409.tlog\FolderStructureCpp.lastbuildstate" wird gelöscht.
InitializeBuildStatus:
  "Release\FolderSt.67F13409.tlog\unsuccessfulbuild" wird erstellt, da "AlwaysCreate" angegeben wurde.
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\CL.exe /c /IIncludes /Zi /nologo /W4 /WX- /diagnostics:classic /sdl /O2 /Oi /Oy- /GL /D CODE_ANALYSIS /D _MBCS /Gm- /EHsc /MD /GS /Gy /fp:precise /permissive- /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"Release\\" /Fd"Release\vc141.pdb" /Gd /TP /analyze /analyze:quiet /analyze:plugin"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\EspXEngine.dll" /analyze:plugin"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\localespc.dll" /FC /errorReport:queue  /analyze:ruleset"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\Rule Sets\NativeRecommendedRules.ruleset" Sources/main.cpp Sources/TestClass.cpp
  main.cpp
e:\dev\source\repos\sonar-cxx-samples\folderstructurecpp\sources\main.cpp(3): warning C4100: "argv": Unreferenzierter formaler Parameter [E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj]
e:\dev\source\repos\sonar-cxx-samples\folderstructurecpp\sources\main.cpp(3): warning C4100: "argc": Unreferenzierter formaler Parameter [E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj]
  TestClass.cpp
  Kompilieren...
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\link.exe /ERRORREPORT:QUEUE /OUT:"E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\Release\FolderStructureCpp.exe" /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG:FULL /PDB:"E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\Release\FolderStructureCpp.pdb" /OPT:REF /OPT:ICF /LTCG:incremental /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\Release\FolderStructureCpp.lib" /MACHINE:X86 /SAFESEH Release\main.obj
  Release\TestClass.obj
  Code wird generiert.
  All 2 functions were compiled because no usable IPDB/IOBJ from previous compilation was found.
  Codegenerierung ist abgeschlossen.
  FolderStructureCpp.vcxproj -> E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\Release\FolderStructureCpp.exe
FinalizeBuildStatus:
  Die Datei "Release\FolderSt.67F13409.tlog\unsuccessfulbuild" wird gelöscht.
  Aktualisieren des Timestamps von "Release\FolderSt.67F13409.tlog\FolderStructureCpp.lastbuildstate".
Die Erstellung von Projekt "E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj" ist abgeschlossen (rebuild Ziel(e)).


Der Buildvorgang wurde erfolgreich ausgeführt.

"E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj" (rebuild Ziel) (1) ->
(ClCompile Ziel) -> 
  e:\dev\source\repos\sonar-cxx-samples\folderstructurecpp\sources\main.cpp(3): warning C4100: "argv": Unreferenzierter formaler Parameter [E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj]
  e:\dev\source\repos\sonar-cxx-samples\folderstructurecpp\sources\main.cpp(3): warning C4100: "argc": Unreferenzierter formaler Parameter [E:\Dev\Source\Repos\sonar-cxx-samples\FolderStructureCpp\FolderStructureCpp.vcxproj]

    2 Warnung(en)
    0 Fehler

Verstrichene Zeit 00:00:03.43
```

### Related information
* Windows version: 10 x64 (1809)
* cxx plugin version: V1.2.1
* SonarQube (in Docker) version:  7.4.0.18908
* Visual Studio version: 2017 Enterprise

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -145,7 +145,7 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
       realPath = absolutePath.toRealPath(LinkOption.NOFOLLOW_LINKS);
     } catch (IOException e) {
       if (LOG.isDebugEnabled()) {
-        LOG.debug("Unable to get the real path: module '{}', baseDir '{}', path '{}', excepiton '{}'",
+        LOG.debug("Unable to get the real path: module '{}', baseDir '{}', path '{}', exception '{}'",
             sensorContext.module().key(), sensorContext.fileSystem().baseDir(), path, e);
       }
       return null;

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -151,10 +151,10 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
       return null;
     }
 
-    // if the real path is equals to the given one - skip search; we already
+    // if the real path is equal to the given one - skip search; we already
     // tried such path
     //
-    // IMPORTANT: don't use Path::equals(), since it's dependent on file-system
+    // IMPORTANT: don't use Path::equals(), since it's dependent on a file-system.
     // SonarQube plugin API works with string paths, so the equality of strings
     // is important
     final String realPathString = realPath.toString();

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -150,8 +150,20 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
       }
       return null;
     }
+
+    // if the real path is equals to the given one - skip search; we already
+    // tried such path
+    //
+    // IMPORTANT: don't use Path::equals(), since it's dependent on file-system
+    // SonarQube plugin API works with string paths, so the equality of strings
+    // is important
+    final String realPathString = realPath.toString();
+    if (absolutePath.toString().equals(realPathString)) {
+      return null;
+    }
+
     return sensorContext.fileSystem()
-        .inputFile(sensorContext.fileSystem().predicates().hasAbsolutePath(realPath.toString()));
+        .inputFile(sensorContext.fileSystem().predicates().hasAbsolutePath(realPathString));
   }
 
   public InputFile getInputFileIfInProject(SensorContext sensorContext, String path) {

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -20,6 +20,9 @@
 package org.sonar.cxx.sensors.utils;
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -135,12 +138,56 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
     }
   }
 
+  private InputFile getInputFileTryRealPath(SensorContext sensorContext, String path) {
+    final Path absolutePath = sensorContext.fileSystem().baseDir().toPath().resolve(path);
+    Path realPath;
+    try {
+      realPath = absolutePath.toRealPath(LinkOption.NOFOLLOW_LINKS);
+    } catch (IOException e) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Unable to get the real path: module {}, baseDir {}, path {}", sensorContext.module().key(),
+            sensorContext.fileSystem().baseDir(), path);
+      }
+      return null;
+    }
+    return sensorContext.fileSystem()
+        .inputFile(sensorContext.fileSystem().predicates().hasAbsolutePath(realPath.toString()));
+  }
+
   public InputFile getInputFileIfInProject(SensorContext sensorContext, String path) {
     if (notFoundFiles.contains(path)) {
       return null;
     }
-    final InputFile inputFile = sensorContext.fileSystem().inputFile(sensorContext.
-      fileSystem().predicates().hasPath(path));
+
+    // 1. try the most generic search predicate first; usually it's the right
+    // one
+    InputFile inputFile = sensorContext.fileSystem()
+        .inputFile(sensorContext.fileSystem().predicates().hasPath(path));
+
+    // 2. if there was nothing found, try to normalize the path by means of
+    // Path::toRealPath(). This helps if some 3rd party tools obfuscate the
+    // paths. E.g. the MS VC compiler tends to transform file paths to the lower
+    // case in its logs.
+    //
+    // IMPORTANT: SQ plugin API allows creation of NewIssue only on locations,
+    // which belong to the module. This internal check is performed by means
+    // of comparison of the paths. The paths which are managed by the framework
+    // (the reference paths) are NOT stored in the canonical form.
+    // E.g. the plugin API neither resolves symbolic links nor performs
+    // case-insensitive path normalization (could be relevant on Windows)
+    //
+    // Normalization by means of File::getCanonicalFile() or Path::toRealPath()
+    // can produce paths, which don't pass the mentioned check. E.g. resolution
+    // of symbolic links or letter case transformation
+    // might lead to the paths, which don't belong to the module's base
+    // directory (at least not in terms of parent-child semantic). This is the
+    // reason why we should avoid the resolution of symbolic links and not use
+    // the Path::toRealPath() as the only search predicate.
+
+    if (inputFile == null) {
+      inputFile = getInputFileTryRealPath(sensorContext, path);
+    }
+
     if (inputFile == null) {
       LOG.warn("Cannot find the file '{}' in module '{}' base dir '{}', skipping violations.",
         path, sensorContext.module().key(), sensorContext.fileSystem().baseDir());

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -145,8 +145,8 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
       realPath = absolutePath.toRealPath(LinkOption.NOFOLLOW_LINKS);
     } catch (IOException e) {
       if (LOG.isDebugEnabled()) {
-        LOG.debug("Unable to get the real path: module {}, baseDir {}, path {}", sensorContext.module().key(),
-            sensorContext.fileSystem().baseDir(), path);
+        LOG.debug("Unable to get the real path: module '{}', baseDir '{}', path '{}', excepiton '{}'",
+            sensorContext.module().key(), sensorContext.fileSystem().baseDir(), path, e);
       }
       return null;
     }

